beater/java_attacher/java_attacher.go

```diff
@@ -24,58 +24,419 @@ import (
+	enabled              bool
+	discoveryRules       []discoveryRule
+	rawDiscoveryRules    []map[string]string
+	agentConfigs         map[string]string
+	downloadAgentVersion string
+	javaBin              string
+	jvmCache             map[int]*jvmDetails
+}
+
+func New(cfg config.JavaAttacherConfig) (*JavaAttacher, error) {
+	logger := logp.NewLogger("java-attacher")
+	if _, err := os.Stat(javaAttacher); err != nil {
+		return nil, err
+	}
+	attacher := &JavaAttacher{
+		logger:               logger,
+		enabled:              cfg.Enabled,
+		agentConfigs:         cfg.Config,
+		downloadAgentVersion: cfg.DownloadAgentVersion,
+		rawDiscoveryRules:    cfg.DiscoveryRules,
+		javaBin:              cfg.JavaBin,
+		jvmCache:             make(map[int]*jvmDetails),
+	}
+	for _, flag := range cfg.DiscoveryRules {
+		for name, value := range flag {
+			switch name {
+			case "include-all":
+				attacher.addDiscoveryRule(includeAllRule{})
+			case "include-user":
+				attacher.addUserDiscoveryRule(value, true)
+			case "exclude-user":
+				attacher.addUserDiscoveryRule(value, false)
+			case "include-main":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-main")
+			case "exclude-main":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			case "include-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, true, "include-vmarg")
+			case "exclude-vmarg":
+				attacher.addCmdLineDiscoveryRule(value, false, "exclude-main")
+			default:
+				logger.Warnf("Ignoring unknown discovery rule %q", name)
+			}
+		}
+	}
+	return attacher, nil
 }
 
-func New(cfg config.JavaAttacherConfig) (JavaAttacher, error) {
-	if cfg.JavaBin == "" {
+func (j *JavaAttacher) addUserDiscoveryRule(user string, isIncludeRule bool) {
+	j.addDiscoveryRule(&userDiscoveryRule{user: user, isIncludeRule: isIncludeRule})
+}
+
+func (j *JavaAttacher) addCmdLineDiscoveryRule(regexS string, isIncludeRule bool, argumentName string) {
+	regex, err := regexp.Compile(regexS)
+	if err != nil {
+		j.logger.Errorf("invalid regex for the %q argument: %v", argumentName, err)
+		return
+	}
+	j.addDiscoveryRule(&cmdLineDiscoveryRule{
+		regex:         regex,
+		isIncludeRule: isIncludeRule,
+		argumentName:  argumentName,
+	})
+}
+
+func (j *JavaAttacher) addDiscoveryRule(rule discoveryRule) {
+	j.discoveryRules = append(j.discoveryRules, rule)
+	j.logger.Debugf("added discovery rule: %s", rule)
+}
+
+func (j *JavaAttacher) findFirstMatch(jvm *jvmDetails) discoveryRule {
+	for _, rule := range j.discoveryRules {
+		if rule.match(jvm) {
+			return rule
+		}
+	}
+	return nil
+}
+
+func (j *JavaAttacher) Run(ctx context.Context) error {
+	if !j.enabled {
+		return fmt.Errorf("java attacher is disabled")
+	}
+
+	if runtime.GOOS == "windows" {
+		// On Windows, we run the attacher by a JVM available to the
+		// current user in continuous mode.
+		if err := j.discoverJavaExecutable(); err != nil {
+			return err
+		}
+		return j.runAttacherContinuous(ctx)
```

Comment:
I think we don't have to do this now that we have support for JVM discovery on Windows. Whenever we can attach to a specific process with the actual Java that it uses, it should be a better option.

Reply:
Clarifying - the attacher does not support running as a different user anyway, but even if it did, it's better to provide it with the user details and a PID rather than start it with the incorrect JVM in continuous mode and let it do all the heavy discovery